extensions/cornerstone-dynamic-volume/src/panels/PanelGenerateImage.tsx

```diff
@@ -0,0 +1,414 @@
+    computedVolume.loadStatus.loaded = true;
+
+    // Set computed scalar data to volume
+    const scalarData = computedVolume.getScalarData();
+    for (let i = 0; i < dataInTime.length; i++) {
+      scalarData[i] = dataInTime[i];
+    }
+
+    // If computed display set does not exist, create an object to be used as
+    // the displaySet. If it does exist, update the image data and vtkTexture
+    if (!computedDisplaySet) {
+      const obj = {
+        [uuidComputedVolume]: {
+          volumeLoaderSchema: computedVolume.volumeId.split(':')[0],
+          displaySetInstanceUID: uuidComputedVolume,
+          SOPClassHandlerId:
+            '@ohif/extension-default.sopClassHandlerModule.stack',
+          SOPClassUID: '1.2.840.10008.5.1.4.1.1.128',
+          SeriesInstanceUID: 'test',
+          StudyInstanceUID: 'test',
+          SeriesNumber: 0,
+          FrameRate: undefined,
+          SeriesDescription: '',
+          Modality: myDynamicVolume.metadata.Modality,
+          isMultiFrame: false,
+          countIcon: undefined,
+          numImageFrames: 1,
+          uid: uuidComputedVolume,
+          averageSpacingBetweenFrames: null,
+        },
+      };
+      setComputedDisplaySet(obj);
+    } else {
+      // update vtkOpenGLTexture and imageData of computed volume
+      const { imageData, vtkOpenGLTexture } = computedVolume;
+      const numSlices = imageData.getDimensions()[2];
+      const slicesToUpdate = [...Array(numSlices).keys()];
+      slicesToUpdate.forEach(i => {
+        vtkOpenGLTexture.setUpdatedFrame(i);
+      });
+      imageData.modified();
+
+      // Check if viewport is currently displaying the computed volume, if so,
+      // call render on the viewports to update the image, if not, call
+      // renderGeneratedImage
+      if (!cache.getVolume(dynamicVolumeId)) {
+        for (const viewport of viewports) {
+          const renderingEngine = cornerstoneViewportService.getRenderingEngine();
+          const viewportForRendering = renderingEngine.getViewport(
+            viewport.viewportId
+          );
+          viewportForRendering.render();
+        }
+      } else {
+        renderGeneratedImage(computedDisplaySet);
+      }
+    }
+  }
+
+  // if computedDisplaySet is defined, render when the data changes
+  useEffect(() => {
+    if (computedDisplaySet) {
+      renderGeneratedImage(computedDisplaySet);
+    }
+  }, [computedDisplaySet]);
+
+  function renderGeneratedImage(displaySet) {
+    commandsManager.runCommand('setDisplaySetToViewport', {
+      displaySet,
+    });
+  }
+
+  function returnTo4D() {
+    const displaySets = displaySetService.getDisplaySetCache();
+    for (const [key, value] of displaySets) {
+      if (value.volumeLoaderSchema === volumeLoaderScheme) {
+        const updatedViewports = hangingProtocolService.getViewportsRequireUpdate(
+          0,
+          key
+        );
+        viewportGridService.setDisplaySetsForViewports(updatedViewports);
+      }
```

Comment:
I guess this is assuming there is only one 4D in the DS. Instead of this you should identify the DSUID, and set it on a ref or something and here just do `getDisplaySetByUID` and do what you do

Reply:
Created ref for DSUID, using that for updatedViewports. I don't think there is a need to call getDisplaySetByUID.